influxql/query_manager.go

```diff
@@ -0,0 +1,156 @@
+package influxql
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/influxdata/influxdb/models"
+)
+
+var (
+	ErrNoQueryManager = errors.New("no query manager available")
+)
+
+// QueryTaskInfo holds information about a currently running query.
+type QueryTaskInfo struct {
+	ID       uint64
+	Query    string
+	Database string
+	Duration time.Duration
+}
+
+// QueryParams holds the parameters used for creating a new query.
+type QueryParams struct {
+	// The query to be tracked. Required.
+	Query *Query
+
+	// The database this query is being run in. Required.
+	Database string
+
+	// The channel to watch for when this query is finished.
+	// Not required, but highly recommended. If this channel is not
+	// set, the query needs to be manually managed by the caller.
+	DetachCh chan struct{}
+}
+
+type QueryManager interface {
+	// AttachQuery attaches a running query to be managed by the query manager.
+	// Returns the query id of the newly attached query or an error if it was
+	// unable to assign a query id or attach the query to the query manager.
+	// This function also returns a channel that will be closed when this
+	// query finishes running.
+	//
+	// After a query finishes running, the system is free to reuse a query id.
+	AttachQuery(params *QueryParams) (qid uint64, closing chan struct{}, err error)
+
+	// DetachQuery removes a query from the query manager. If a DetachCh is used
+	// to attach the query, this method will be called automatically when the
+	// DetachCh is closed. This method can also be used to forcefully terminate
+	// a running query.
+	DetachQuery(qid uint64) error
+
+	// ListQueries lists the currently running tasks.
+	ListQueries() []QueryTaskInfo
```

Comment:
`List` doesn't seem necessary. Just `Queries()` is more concise and understandable.